sdks/java/extensions/sql/src/main/java/org/apache/beam/sdk/extensions/sql/SqlTransform.java

```diff
@@ -154,10 +165,21 @@ public static SqlTransform query(String queryString) {
         .setQueryString(queryString)
         .setUdafDefinitions(Collections.emptyList())
         .setUdfDefinitions(Collections.emptyList())
+        .setTableProviderMap(Collections.emptyMap())
         .setAutoUdfUdafLoad(false)
         .build();
   }
 
+  public SqlTransform withTableProvider(String name, TableProvider tableProvider) {
```

Comment:
This introduces a new way to register table providers, and I am thinking if it's needed to be introduced into BeamSQL right now? 

The auto service way is already good enough for in-memory table provider and external table provider.

Reply:
Currently we have 2 ways to use SQL:
  - load all table providers, register them with top-level `beam` schema;
  - don't load anything, only register `PCollection` table providers in `beam` schema for current pipeline;

This adds a third way:
  - register an alternative top-level schema that allows you to join `PCollection` with things coming from metastore.

Reply:
I am thinking  if we need to introduce registering approach.

For existing auto service loader approach, if we add annotation with a parameter schema, will that be efficiently to register table providers to different schema during initialization step?

Reply:
I may not understand the details. But if something autoservice is on the classpath and registers, say `hive` schema with databases and tables, is it available to the `SqlTransform` without calling this builder method?

Reply:
In general I am not opposed to the idea of auto-registering. But I don't see it as an alternative to this approach, it seems to be a separate use case.

That is, to register the schemas/table providers users still need to do it explicitly. Either in a mapping somewhere (either static on a class, or externally like a module in guice), or a configuration code close to the call site. 

Simple static mapping with annotations doesn't work, you can potentially have multiple instances of the same table provider configured to access different back-ends (imagine 2 hadoop clusters with different metastores).

Dynamic mapping from name to a table provider instance separately from the pipeline code as part of development/configuration is fine but it doesn't solve us anything at the moment. Users still have to write the mapping, and our code needs to understand where to read the mapping from. In case if it's not right next to the calling code it makes things less obvious and harder to debug. Such approach would work well if we, for example, needed to replace the mappings often (e.g. imagine plugins, replacing IOs in templates), or if the amount and configuration of table providers would make it cleaner to do separately, maybe in other cases. However I don't see an immediate need for this.

What the approach in this PR solves is ability to explicitly add a couple of extra sources inline, next to where they are used. I think we can get immediate benefits from this without blocking any alternative approaches.And the syntax is somewhat similar to how IOs, windows and other things are applied in Beam at the moment.

Reply:
@kennknowles 

It will be only available in `beam` schema (which is the default) without calling this builder method, assuming that hive source implements current table provider.

Reply:
To ease the pain on understanding the auto service approach and manually registration approach. I would like to propose we can limit the auto service approach to only load in-memory table provider into `beam` schema, and then any other schema registration have to be registered manually. In order to enforce it, we can start from separating table provider interfaces. The auto service can only load in-memory table provider and external sources should use external table provider interface.